docs/user/testing.md

```diff
@@ -0,0 +1,189 @@
+
+In garage, we use [pytest](https://docs.pytest.org/en/stable/getting-started.html) to carry out tests.
+
+## Test Modules
+
+All test files are stored under `tests/` in the garage respository.
+
+Test modules are organized in the same way as the garage main repository. Ideally, all main modules and files under `src/garage/` should be covered with test cases.
+
+Test modules of garage are structured in the following ways:
+
+- `tests/` contains all tests and supporting code
+- `tests/garage` contains unit tests
+- `tests/integration_tests` contains integration tests
+- modules in `tests/` and `tests/fixtures` contain helper codes and fixtures which make writing good tests easier.
+
+```bash
+tests
+├── fixtures
+├── garage
+├── integration_tests
+├── helpers.py
+└── (helper_files.py)
+```
+
+## Prerequisites
+
+To begin testing, we suggest using the same Python manager environment i.e. `virtualenv`, `conda` as we develop on garage repository so that all packages and dependencies required for testing are installed in the virutal environment.
+
+Let's begin by activating the virtual environment:
+
+```bash
+# Conda
+$ conda activate myenv
+
+# Virtualenv
+$ source myenv/bin/activate
+```
+
+Next, we need to install pytest. Generally, pytest should have already been installed upon garage's `dev` installation. To install the garage environment for testing, you will need to install the `garage[all,dev]` for dependencies such as pytest, mujoco and so on.
+
+Also, you may want to check out our installation guide first before diving into tests. It is also recommened to check out official [pytest](https://docs.pytest.org/en/stable/getting-started.html) documentation.
+
+```bash
+$ cd path/to/garage/
+$ pip install -e '.[all,dev]'
+```
+
+To get started, run pytest as follows:
+
+```bash
+$ pytest
+```
+
+## Basic Pytest Usage
+
+Congrats! Now you're ready for testing garage!
+
+We will start with a simple test. To be consistent with pytest requirement and garage modules, we name our test `.py` file with pre-fix `test_`.
+
+Let's write a simple test case for the vpg algorithm `src/garage/tf/algos/vpg.py`.
+
+We begin by creating a file called `test_vpg.py` and put it under the tree `tests/garage/tf/algos/`.
+
+We want to test VPG in the cart pole environment. Hence, we create a unitest named `test_vpg_cartpole`. Inside the function, we define the environemnt, the policy and the baselien and feed them to the VPG algorithm and run the experiment. Finally, we make sure the return value is identical to our expectation by using `assert`.
+
+In new xunit-style tests, multiple tests are modeled into a class for modular and scalable structure. There is no need to subclass or anything, but make sure the prefix of the class starts with `Test`, otherwise the class will be skipped. Note that it's not encouraged to use this style, especially when a test doesn't require `setup_method` and `teardown_method`.
```

Comment:
To me, the first sentence sounds like that's the recommended way, while it's not.

Things I think that need to be mentioned are tests can be top level functions, don't subclass unittest.TestCase and basically follow pytest's recommendation.